api/src/getStartDate.ts

```diff
@@ -0,0 +1,144 @@
+        let regexp = /(\d{2})\/(\d{2})\/(\d{4})/;
+        let day, month, year;
+        let matched = termStartDate.match(regexp);
+        if (matched != null) {
+            year = matched[3];
+            month = matched[2];
+            day = matched[1];
+        } else {
+            res.send("Error getting the start date");
+        }
+        const termStartDateObj = new Date(year, month, day);
+        // Getting the difference in time so Freerooms gets the most recent term start date when
+        // it starts.
+        const timeDelta = currDate.valueOf() - termStartDateObj.valueOf();
+
+        if (timeDelta >= 0) {
+            // If the data is more than a day old, we will return the term start date that is for the
+            // new term, else we get the date for the old term.
+            res.send(termStartDate);
+        } else {
+            let dummyTerm = termArray.findIndex((term) => term == currTerm) - 1;
+            // Checking if the term is the first term, we can have case when index out of bounds.
+            if (dummyTerm < 0) {
+                dummyTerm = 0; // Sanity check to make sure the array does not break
+                // Only time this will happen is when the next year's summer starts over.
+            }
+            const oldTermStartDate =
+                data.timetableData[termArray[dummyTerm]][FIRST_COURSE]["classes"][0]["termDates"][
+                    "start"
+                ];
+            res.send(oldTermStartDate);
+        }
+    } catch (e) {
+        res.status(400).send("Error");
+    }
+};
+
+/**
+ * Get the latest term start date from data.json
+ */
+const getStartDateNotangles = (req: express.Request, res: express.Response) => {
+    try {
+        // Get the latest term and check the first class of the term
+        // The assumption is that the starting date is the start date of the first class
+        // of the term.
+        const termStartDate =
+            data.timetableData[term()][FIRST_COURSE]["classes"][0]["termDates"]["start"];
+        res.send(termStartDate);
+    } catch (e) {
+        res.status(400).send("Error");
+    }
+};
+
+/**
+ * Get the latest term start date from data.json
+ */
+const getAvailableTermData = (req: express.Request, res: express.Response) => {
+    try {
+        res.send(term());
+    } catch (e) {
+        res.status(400).send("Error");
+    }
+};
+
+/**
+ * Get the term which has data from data.json, this term
+ * is the latest term with data. This is exclusively for Notangles, so
+ * Freerooms please do not use it!
+ * @returns the latest term number
+ */
+const term = () => {
+    try {
+        let term: string = "";
+        for (let termId of termArray) {
+            let timetableData = data.timetableData;
+            // Check if the termId exists in the timetableData and the fields are actually
+            // present. If they are, return the term as it is.
+            if (timetableData.hasOwnProperty(termId)) {
```

Comment:
I think if you reversed the terms in the `termArray` (i.e. start from T3 and go backwards) you can have an early return here and simplify your code. The `term` variable would also be unnecessary I think. Let me know if you can't figure out what I mean. I don't think the reversal would affect the other places where `termArray` is used.